juju/model.py

```diff
@@ -0,0 +1,210 @@
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+    def _lazy_load_peer_units(self):
+        # In lieu of an implicit peer relation, goal-state seems to be the only way to find out
+        # what our peer units are. The information it returns is rather limited, but is enough
+        # for our purposes here.
+        return [Unit(unit_name) for unit_name in self._backend.goal_state()['units']]
+
+
+class Application:
+    def __init__(self, name, units):
+        self.name = name
+        self._units = units
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+    @property
+    def units(self):
+        # This property allows for the list of units to be lazy-loaded by passing in a function instead of a list.
+        if callable(self._units):
+            self._units = self._units()
+        return self._units
+
+    def __hash__(self):
+        # An Application instance from one relation will not be identical to one from another relation,
+        # since they may have different views of the units of that app. However, they should be considered
+        # equivalent, since they are views of the same entity. Generally, this should only matter for using
+        # model.app as an index into a given relation's relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Application):
+            return False
+        return self.name == other.name
+
+
+class Unit:
+    def __init__(self, name):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            return False
+        return self.name == other.name
+
+
+class RelationMap(Mapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, local_app, local_unit, backend=None):
+        self._relation_names = relation_names
+        self._local_app = local_app
+        self._local_unit = local_unit
+        self._backend = backend or ModelBackend()
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __iter__(self):
+        return iter(self._relation_names)
+
+    def __len__(self):
+        return len(self._relation_names)
+
+    @lru_cache(maxsize=None)
+    def __getitem__(self, relation_name):
+        # Lazy-load the list of relations for this relation name.
```

Comment:
The comment may be dropped now. Whether it's lazy loaded or not is now a job for the backend to decide.

Reply:
Technically, the only way for it to not be lazy-loaded (or loaded on demand) would be for it to not use a backend at all and to be passed in all the data up front during `__init__`.  Also, the caching decision is being made here (via `@lru_cache`) and not the backend because doing caching in the backend would lead to more duplicates of the model class instances (`Application`, `Unit`, `Relation`, etc.) being created, at least unless we add de-duplication logic as mentioned below.  And caching closer to the result that will actually be used seems more efficient.